Bluetooth: Classic: HFP_AG: Improve the SCO management

lylezhu2012 opened the pull request:
* Bluetooth: Classic: HFP_AG: Improve the SCO management

There are two issues found,
Issue 1, if the SCO connection has been established, the SCO connected event will not be notified. Then the call that waiting for the SCO connected event will not be updated. And the call will be rejected due to the timeout.
Issue 2, the SCO con object will be cleared when SLC is disconnecting and the SCO connection has been established. It causes the SCO connect object will be `NULL` in the SCO disconnected callback.

Add a new field `sco_conn` to keep the SCO connect object. Manage the SCO connection via the field `ag->sco_conn`.

When creating the SCO connection, keep the created SCO connect object to `ag->sco_conn`. And if the `ag->sco_chan.sco` is `NULL`, it means the SCO connection has been established. Just update the call that waiting for the SCO connected event directly.
Increase the `ref` of SCO connect object and pass to the field `ag->sco_conn` if the `ag->sco_conn` is `NULL`.
Decrease the `ref` of SCO connect object if the `ag->sco_conn` is not `NULL` when disconnect the SCO connection or process SCO disconnected event.

* Bluetooth: Classic: SCO: Add SCO conn connected/disconnected cb

Add two functions to register/unregister SCO conn connected/disconnected callback.

Add macro BT_SCO_CONN_CB_DEFINE to defined static SCO connect callback. Define section `bt_sco_conn_cb` to ROM.

* Bluetooth: Classic: HFP_AG: Register SCO connect callback

There is an issue that the SCO connect cannot be un-referenced by HFP AG when the SCO connection is broken if the SCO connect is not created by HFP AG.

Register SCO connect change callback. And un-reference the SCO connect in SCO disconnected callback.

lylezhu2012 committed:
Bluetooth: Classic: SCO: Add SCO conn connected/disconnected cb

Add two functions to register/unregister SCO conn connected/
disconnected callback.

Add macro BT_SCO_CONN_CB_DEFINE to defined static SCO connect callback.
Define section `bt_sco_conn_cb` to ROM.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Classic: HFP_AG: Register SCO connect callback

There is an issue that the SCO connect cannot be un-referenced by HFP
AG when the SCO connection is broken if the SCO connect is not created
by HFP AG.

Register SCO connect change callback. And un-reference the SCO connect
in SCO disconnected callback.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -280,7 +280,11 @@ static struct bt_hfp_ag_call *get_new_call(struct bt_hfp_ag *ag, const char *num
 			/* Copy number to ag->number including null-character */
-			strcpy(call->number, number);
+			if (number != NULL) {
+				strcpy(call->number, number);
```
The function `get_new_call()` is called in 5 places. In function `bt_hfp_ag_remote_incoming()` and `bt_hfp_ag_outgoing_call()`, the length of `number` has been checked before calling the function `get_new_call()`. For other new places, adding in this PR, the passed buffer is an array `char number[CONFIG_BT_HFP_AG_PHONE_NUMBER_MAX_LEN + 1];` defined in structure `struct bt_hfp_ag_ongoing_call`.

I think there is an issue here. I need to add comments for `char number[CONFIG_BT_HFP_AG_PHONE_NUMBER_MAX_LEN + 1];` to make sure the last character is '\0'.

lylezhu2012 replied:
Updated the doxygen comments.
And checked the string length of `bt_hfp_ag_ongoing_call::number` before calling the function `get_new_call()`.

jhedberg commented on subsys/bluetooth/host/classic/hfp_internal.h:
```diff
@@ -136,6 +136,12 @@
 /* Call held by Response and Hold */
 #define BT_HFP_CLCC_STATUS_CALL_HELD_HOLD 6
 
+/* CLCC direction */
+/* CLCC direction: incoming */
+#define BT_HFP_CLCC_DIR_INCOMING 1
+/* CLCC direction: outgoing */
+#define BT_HFP_CLCC_DIR_OUTGOING 0
```
Right, but if you define them identically, then do you really need the duplicate definitions? You're even using `__packed` for the enum, meaning you have some special protocol requirements for it.

lylezhu2012 replied:
> Right, but if you define them identically, then do you really need the duplicate definitions? You're even using `__packed` for the enum, meaning you have some special protocol requirements for it.

I think from the value perspective, they are can be shared.

gzh-terry replied:
I think it depends on whether these definitions have the same source. For example, if `BT_HFP_AG_CALL_DIR_INCOMING` is defined by Zephyr and is aimed to have a distinct layer that isolates the stack and upper layer, it should be defined with a different name.

Without this goal, I would recommend having the same definition. Assuming that both definitions come from 3GPP
![image](https://github.com/user-attachments/assets/3c9d8e53-da98-497b-baf0-0a32cdf471cf)

lylezhu2012 replied:
The initial idea was to use `bool` variable to mark whether it is an incoming or outgoing call. But the `bool` type is not intuitive, so two values ​​are defined. So from my point of view, these two macros are not part of the 3GPP protocol.

lylezhu2012 commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -1054,20 +1201,455 @@ static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+	hfp_ag_lock(ag);
+	call_state = call->call_state;
+	hfp_ag_unlock(ag);
+	if ((call_state == BT_HFP_CALL_INCOMING) || (call_state == BT_HFP_CALL_OUTGOING)) {
```
I remember there was a rule in MISRA that required explicit parentheses.

lylezhu2012 committed:
Bluetooth: Classic: HFP_AG: Improve the SCO management

There are two issues found,
Issue 1, if the SCO connection has been established, the SCO connected
event will not be notified. Then the call that waiting for the SCO
connected event will not be updated. And the call will be rejected due
to the timeout.
Issue 2, the SCO con object will be cleared when SLC is disconnecting
and the SCO connection has been established. It causes the SCO connect
object will be `NULL` in the SCO disconnected callback.

Add a new field `sco_conn` to keep the SCO connect object. Manage the
SCO connection via the field `ag->sco_conn`.

When creating the SCO connection, keep the created SCO connect object
to `ag->sco_conn`. And if the `ag->sco_chan.sco` is `NULL`, it means
the SCO connection has been established. Just update the call that
waiting for the SCO connected event directly.
Increase the `ref` of SCO connect object and pass to the field
`ag->sco_conn` if the `ag->sco_conn` is `NULL`.
Decrease the `ref` of SCO connect object if the `ag->sco_conn` is not
`NULL` when disconnect the SCO connection or process SCO disconnected
event.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Classic: SCO: Modify `bt_sco_chan::ops` with `const`

Use `const` to modify the field `ops` of the structure
`struct bt_sco_chan`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/hfp_hf.c:
```diff
@@ -592,6 +592,43 @@ static void hf_call_state_update(struct bt_hfp_hf_call *call, int state)
+			continue;
+		}
+
+		count++;
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -1054,20 +1201,455 @@ static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+	atomic_set_bit_to(call->flags, BT_HFP_AG_CALL_INCOMING, dir == BT_HFP_AG_CALL_DIR_INCOMING);
+
+	if (call_count > 1) {
+		if (dir) {
```
Fixed.

lylezhu2012 commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -1054,20 +1201,455 @@ static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+			continue;
+		}
+
+		if (clear) {
+			if (atomic_test_and_set_bit(call->flags, bit)) {
+				return call;
+			}
```
Add `get_call_clear_flag()` for `clear` is `true` case.

lylezhu2012 commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -1054,20 +1201,455 @@ static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+	if ((call_state == BT_HFP_CALL_INCOMING) ||
+		atomic_test_and_clear_bit(call->flags, BT_HFP_AG_CALL_ALERTING)) {
+		bt_hfp_ag_set_call_state(call, BT_HFP_CALL_ALERTING);
+		bt_hfp_ag_call_ringing_cb(call, true);
```
Removed the `hfp_ag_lock/unlock` for cases that the field of AG is read-only accessed.

lylezhu2012 commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -1054,20 +1201,455 @@ static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+
+static struct bt_conn *bt_hfp_ag_create_sco(struct bt_hfp_ag *ag)
+{
+	static struct bt_sco_chan_ops ops = {
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -3722,6 +4067,31 @@ static int bt_hfp_ag_sco_accept(const struct bt_sco_accept_info *info,
+
+	ARRAY_FOR_EACH(bt_hfp_ag_pool, i) {
+		if (bt_hfp_ag_pool[i].sco_conn == conn) {
+			if (bt_hfp_ag_pool[i].sco_conn != NULL) {
```
Fixed.

jhedberg commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -1055,19 +1153,452 @@ static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+			continue;
+		}
+
+		if (clear) {
+			if (atomic_test_and_set_bit(call->flags, bit)) {
```
Shouldn't this be `test_and_clear_bit()`?

lylezhu2012 replied:
Yes. You are correct.

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -1055,19 +1153,452 @@ static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+	return NULL;
+}
+
+#define get_call_with_flag(ag, bit) _get_call_with_flag(ag, bit, false)
+#define get_call_clear_flag(ag, bit) _get_call_with_flag(ag, bit, true)
```
Prefer actual static functions instead of macros. The resulting binary will be exactly the same but you get the added benefit of type enforcement from the compiler. You don't even need to use `inline` since the compiler will automatically do it for you.

lylezhu2012 replied:
Updated them according to the following comments.

jhedberg commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -1055,19 +1153,452 @@ static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+	}
+}
+
+static struct bt_hfp_ag_call *_get_call_with_flag(struct bt_hfp_ag *ag, int bit, bool clear)
```
I'd avoid the `_` prefix. Why can't you implement the "don't clear" version and then the "clear" version first calls the "don't clear" one, then clears the flag before returning, i.e.:

```c
static struct bt_hfp_ag_call *get_call_with_flag(struct bt_hfp_ag *ag, int flag)
{
	...
}

static struct bt_hfp_ag_call *get_call_clear_flag(struct bt_hfp_ag *ag, int flag)
{
	struct bt_hfp_ag_call *call;

	call = get_call_with_flag(ag, flag);
	if (call != NULL) {
		atomic_clear_bit(ag->flags, flag);
	}

	return call;
}
```

lylezhu2012 replied:
Updated the function `get_call_with_flag()` and `get_call_clear_flag()`.

jhedberg commented on subsys/bluetooth/host/classic/sco.c:
```diff
@@ -403,3 +442,27 @@ struct bt_conn *bt_conn_create_sco(const bt_addr_t *peer, struct bt_sco_chan *ch
+
+int bt_sco_conn_cb_unregister(struct bt_sco_conn_cb *cb)
+{
+	CHECKIF(cb == NULL) {
+		return -EINVAL;
+	}
```
Any reason you're doing a NULL check for `unregister()` but not for `register()`?

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/hfp.c:
```diff
@@ -471,6 +484,16 @@ static int cmd_select_codec(const struct shell *sh, size_t argc, char **argv)
+		hf_auto_select_codec = true;
+	} else {
+		hf_auto_select_codec = false;
+	}
```
The argument of the command is `yes` or `no`. Yes, if it is `on/off`, `enable/disable`, or `true/false`, we can leverage the function `shell_strtobool()`. I think it is OK to change the argument of the command to `enable/disable`.

I have updated the PR.

jhedberg replied:
> The argument of the command is `yes` or `no`.

Is there a good reason to be that specific & restrictive? Ultimately it just gets converted into a `bool`, so IMO "true"/"false" would be equally intuitive.

lylezhu2012 replied:
Yes. I have changed to `enable/disable`.
